File: Front-End/src/modules/cases/types/case.ts
import { CASE_STATES, ATTENTION_TYPES, GENDER_OPTIONS } from '@/core/config/api.config'

export type CaseState = typeof CASE_STATES[keyof typeof CASE_STATES]
export type AttentionType = typeof ATTENTION_TYPES[keyof typeof ATTENTION_TYPES]
export type Gender = typeof GENDER_OPTIONS[keyof typeof GENDER_OPTIONS]

export enum CasePriority {
  NORMAL = 'Normal',
  PRIORITARIO = 'Prioritario'
}

export interface EntityInfo {
  id: string
  name: string
}

export interface PatientInfo {
  patient_code: string
  name: string
  age: number
  gender: Gender
  entity_info: EntityInfo
  care_type: AttentionType
  observations?: string
  updated_at?: string
  // Campos en español para compatibilidad
  paciente_code?: string
  nombre?: string
  edad?: number
  sexo?: Gender
  entidad_info?: EntityInfo
  tipo_atencion?: AttentionType
  fecha_actualizacion?: string
}

export interface DoctorInfo {
  nombre: string
}

export interface TestInfo {
  id: string
  name: string
  quantity?: number
  // Campos en español para compatibilidad
  nombre?: string
  cantidad?: number
}

export interface SampleInfo {
  body_region: string
  tests: TestInfo[]
  // Campos en español para compatibilidad
  region_cuerpo?: string
  pruebas?: TestInfo[]
}

export interface PathologistInfo {
  id: string
  name: string
}

export interface CaseResult {
  method?: string[]
  macro_result?: string
  micro_result?: string
  diagnosis?: string
  observations?: string
}

export interface CaseModel {
  id?: string
  case_code: string
  patient_info: PatientInfo
  requesting_physician?: string
  service?: string
  samples: SampleInfo[]
  state: CaseState
  priority?: CasePriority
  created_at: string
  updated_at: string
  observations?: string
  assigned_pathologist?: PathologistInfo
  entity_info?: EntityInfo
  result?: CaseResult
  created_by?: string
  updated_by?: string
  active?: boolean
}

export interface CaseListItem {
  _id: string
  case_code: string
  patient: { name: string; patient_code: string }
  state: CaseState
  created_at: string
  assigned_pathologist?: { name: string }
}

export interface CaseStatistics {
  total_cases: number
  pending_cases: number
  in_progress_cases: number
  completed_cases: number
  delivered_cases: number
  cancelled_cases: number
  expired_cases: number
  cases_without_pathologist: number
}

File: Back-End/app/modules/cases/schemas/case.py
from typing import Optional, List, Dict
from datetime import datetime
from pydantic import BaseModel, Field


class EntityInfo(BaseModel):
    id: str = Field(..., max_length=50)
    name: str = Field(..., max_length=200)


class PatientInfo(BaseModel):
    patient_code: str = Field(..., max_length=50)
    name: str = Field(..., max_length=200)
    age: int = Field(..., ge=0, le=150)
    gender: str = Field(..., max_length=20)
    entity_info: EntityInfo
    care_type: str = Field(..., max_length=50)
    observations: Optional[str] = Field(None, max_length=1000)


class SampleTest(BaseModel):
    id: str = Field(..., max_length=50)
    name: str = Field(..., max_length=200)
    quantity: int = Field(default=1, ge=1)


class SampleInfo(BaseModel):
    body_region: str = Field(..., max_length=100)
    tests: List[SampleTest] = Field(default_factory=list)


class AssignedPathologist(BaseModel):
    id: str = Field(..., max_length=50)
    name: str = Field(..., max_length=200)


class CasePriority(str):
    NORMAL = "Normal"
    PRIORITARIO = "Prioritario"


class CaseState(str):
    EN_PROCESO = "En proceso"
    POR_FIRMAR = "Por firmar"
    POR_ENTREGAR = "Por entregar"
    COMPLETADO = "Completado"


class AdditionalNote(BaseModel):
    date: datetime = Field(..., description="Fecha de la nota")
    note: str = Field(..., max_length=1000, description="Contenido de la nota")
    
    class Config:
        from_attributes = True


class CaseCreate(BaseModel):
    patient_info: PatientInfo
    requesting_physician: Optional[str] = Field(None, max_length=200)
    service: Optional[str] = Field(None, max_length=100)
    samples: Optional[List[SampleInfo]] = Field(default_factory=list)
    state: str = Field(default=CaseState.EN_PROCESO)
    priority: str = Field(default=CasePriority.NORMAL)
    observations: Optional[str] = Field(None, max_length=1000)


class CaseUpdate(BaseModel):
    patient_info: Optional[PatientInfo] = None
    requesting_physician: Optional[str] = Field(None, max_length=200)
    service: Optional[str] = Field(None, max_length=100)
    samples: Optional[List[SampleInfo]] = None
    state: Optional[str] = None
    priority: Optional[str] = None
    observations: Optional[str] = Field(None, max_length=1000)
    assigned_pathologist: Optional[AssignedPathologist] = None
    delivered_to: Optional[str] = Field(None, max_length=200)
    delivered_at: Optional[datetime] = None
    business_days: Optional[int] = Field(None, ge=0, description="Business days elapsed")
    additional_notes: Optional[List[AdditionalNote]] = None


class CaseResult(BaseModel):
    method: Optional[List[str]] = None
    macro_result: Optional[str] = None
    micro_result: Optional[str] = None
    diagnosis: Optional[str] = None
    observations: Optional[str] = None
    cie10_diagnosis: Optional[Dict[str, str]] = None
    cieo_diagnosis: Optional[Dict[str, str]] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class CaseResponse(BaseModel):
    id: str
    case_code: str
    patient_info: PatientInfo
    requesting_physician: Optional[str] = None
    service: Optional[str] = None
    samples: List[SampleInfo] = Field(default_factory=list)
    state: str
    priority: str
    observations: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    signed_at: Optional[datetime] = None
    assigned_pathologist: Optional[AssignedPathologist] = None
    result: Optional[CaseResult] = None
    delivered_to: Optional[str] = None
    delivered_at: Optional[datetime] = None
    business_days: Optional[int] = None
    additional_notes: Optional[List[AdditionalNote]] = None

    class Config:
        from_attributes = True


